### PR TITLE
Don't include vnet name in template type

### LINKED ISSUE #4556: Deploying more than topology based on the default AKS ClusterClass only results in one successfully deployed cluster
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**

- Deployed the default `aks-clusterclass` template
- Deployed three of the `aks-topology` template
- Only one cluster is successfully created

I lost a handle on the exact error, but the error was while creating the managed cluster resource and was indicating that the vnet could not be found. The ASO vnet's default name is the name of the AzureManagedControlPlaneTemplate which is shared among all instances of the template, so subsequent topologies also try to share the same ASO vnet instead of creating their own in their respective resource groups.

**What did you expect to happen:**

Three clusters are deployed successfully.

**Anything else you would like to add:**

This is the smallest change I need to make to get things to work the way I expect:
```diff
diff --git a/api/v1beta1/azuremanagedcontrolplanetemplate_default.go b/api/v1beta1/azuremanagedcontrolplanetemplate_default.go
index fce6a8182..de87e97af 100644
--- a/api/v1beta1/azuremanagedcontrolplanetemplate_default.go
+++ b/api/v1beta1/azuremanagedcontrolplanetemplate_default.go
@@ -38,9 +38,9 @@ func (mcp *AzureManagedControlPlaneTemplate) setDefaults() {
 
 // setDefaultVirtualNetwork sets the default VirtualNetwork for an AzureManagedControlPlaneTemplate.
 func (mcp *AzureManagedControlPlaneTemplate) setDefaultVirtualNetwork() {
-	if mcp.Spec.Template.Spec.VirtualNetwork.Name == "" {
-		mcp.Spec.Template.Spec.VirtualNetwork.Name = mcp.Name
-	}
 	if mcp.Spec.Template.Spec.VirtualNetwork.CIDRBlock == "" {
 		mcp.Spec.Template.Spec.VirtualNetwork.CIDRBlock = defaultAKSVnetCIDR
 	}
```

This raises the question though of how much defaulting we should do on template types when the same defaulting is already being applied to the non-template type.

An alternative approach is to add extra API fields to the resources which define the names of the ASO resources separately from the names of the resources as they exist in Azure. Or to include the resource group name in the names of all ASO resources, but I'm not sure how easily that might hit name length limits or how easily we could maintain backwards compatibility with that.

**Environment:**

- cluster-api-provider-azure version: 40f3dfb2d7bf128a267b33976c7e0e37f9f530eb
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by willie-yao**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
This PR fixes a bug where the `name` field for the virtual network is erroneously being defaulted to a name that is shared across clusters. The vnet name should not be shared across clusters and this will cause issues with any cluster that is created via a ClusterClass template after the first one.

To avoid breaking API changes, this PR does the following:
- Print out a webhook warning if the vnet name is specified in the AzureManagedControlPlaneTemplate
- Update docs and comments to reflect this change

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4556

**Special notes for your reviewer**:

- [x] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Fix ClusterClass MP: Don't include vnet name in template type. BREAKING Go API but not real-world usage
```

### CODE REVIEW

**Comment by nojnhuh** [comment]
I feel pretty strongly that we should not be backporting API breaking changes to release branches, even if it's to fix a bug.

**Comment by willie-yao** [comment]
> I feel pretty strongly that we should not be backporting API breaking changes to release branches, even if it's to fix a bug.

That's fair, is there a way to cancel to cherry-pick PRs or do we just have to manually close them?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanetemplates.yaml
Comment (nojnhuh) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanetemplates.yaml:733]: Are we sure this PR is actually doing what we want? Isn't the idea that the name here should be removed from the template type?

Comment (willie-yao) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanetemplates.yaml:733]: Whoops! This line should refer to the `ClassSpec` type. Will make this change now!

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanes.yaml
Comment (willie-yao) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanes.yaml:781]: Not sure why the `allOf` pops up when adding a separate `ManagedControlPlaneVirtualNetworkClassSpec` type.. This is causing the CRD to be invalid. Will be looking into this.

File: api/v1beta1/types_class.go
Comment (jackfrancis) [inline api/v1beta1/types_class.go:97]: I don't know how we can make these changes without breaking the API contract

Comment (willie-yao) [reply api/v1beta1/types_class.go:97]: Yeah it's tricky. Also doing it this way messes with the CRDs per my comment above. I think I'm misunderstanding how template types work when they are a field within a template type. For example, we want some fields of `VirtualNetwork` to be shared while others (like `name`) shouldn't. This pattern exists for `FleetsMember` but doing so for the vnet doesn't seem to work https://github.com/willie-yao/cluster-api-provider-azure/blob/a3758946dbc5de3b5941100d41b599328e42a892/api/v1beta1/types_class.go#L225

I'll do some more digging into this to figure out a solution!

Comment (willie-yao) [reply api/v1beta1/types_class.go:97]: If we are unable to make an api change here, would it be possible to use webhook validation to block setting `name` if the amcp is being created from a template? It's definitely not ideal.

Comment (nojnhuh) [reply api/v1beta1/types_class.go]: I would still classify making this invalid in the webhook as a breaking change I think, at least in the sense that "I have this YAML that used to work but now doesn't."

I think the most we can do here without breaking anyone is to update our examples to patch in the vnet name per-cluster vs. defining it on the template, then make the webhook return a warning if users do define it in the template to say "you probably don't want to do this."

Comment (willie-yao) [reply api/v1beta1/types_class.go]: Made this change in this commit: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4735/commits/f3f78a757967b5d60b11931a0127f8cf45e21867

It basically overwrites the field if someone sets the vnet name from the template. Is that the right idea?

Comment (nojnhuh) [reply api/v1beta1/types_class.go]: If I'm allowed to split hairs, I would still call that a breaking behavioral change that I think still falls within the contract if a user defines the vnet name in a template and that's no longer honored by CAPZ.

Comment (willie-yao) [reply api/v1beta1/types_class.go]: Ah okay I think I misread your comment. When you say "update our examples" do you just mean the doc examples? And the only thing we should do here is add a webhook warning.

Comment (nojnhuh) [reply api/v1beta1/types_class.go]: The docs at a minimum, but I think we should also update all the templates in order to follow those new docs. A webhook warning still sgtm.

Comment (willie-yao) [reply api/v1beta1/types_class.go]: Gotcha. None of our templates explicitly specify the `virtualNetwork` field in ClusterClass so we're good on that front. I'll update the docs and keep the webhook warning!

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (nojnhuh) [inline api/v1beta1/azuremanagedcontrolplane_types.go:383]: If this ultimately overlaps with what's defined in the CRD already, maybe we don't want to specify the name here?

File: api/v1beta1/types_class.go
Comment (willie-yao) [reply api/v1beta1/types_class.go:418]: I saw your previous comment and realized that I just forgot to revert this since we're still keeping the CRDs the same. The bug fix will be in the template webhook which defaulted the vnet to the same name.

Comment (jackfrancis) [reply api/v1beta1/types_class.go:418]: Case-by-case basis? What's the path that is the easiest to maintain going forward (and easiest for future maintainers to understand the history of how we got here)?

Comment (willie-yao) [reply api/v1beta1/types_class.go:418]: I think that it makes a bit more sense to move the name out of the class spec so that the logic is only incorrect in one place (VirtualNetwork in the template type not being set to the respective vnet template type: https://github.com/willie-yao/cluster-api-provider-azure/blob/11e92174a84cd508f5900a68b135016ef75d56b3/api/v1beta1/types_class.go#L97).

Comment (willie-yao) [reply api/v1beta1/types_class.go:418]: The name should never be shared, so leaving it in the template spec may be a bit confusing.

Comment (nojnhuh) [reply api/v1beta1/types_class.go:418]: That sounds like a good rule of thumb. Here it seems like mostly a wash to me.

Comment (willie-yao) [reply api/v1beta1/types_class.go:418]: > Here it seems like mostly a wash to me.

Do you mean that we should not change the type here? Sorry I'm just unfamiliar with the saying but I think either way works!

Comment (nojnhuh) [reply api/v1beta1/types_class.go:418]: I mean I could go either way here.

Comment (nojnhuh) [reply api/v1beta1/types_class.go:418]: No strong preference

Comment (willie-yao) [reply api/v1beta1/types_class.go:418]: @jackfrancis What is your opinion on moving the `name` out of the template type?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanetemplates.yaml
Comment (jackfrancis) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanetemplates.yaml:733]: Has this change been applied?

Comment (willie-yao) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanetemplates.yaml:733]: No. We realized that this would be a breaking API change. The AzureManagedControlPlaneClassSpec references the incorect non-template type instead of the template type here: https://github.com/willie-yao/cluster-api-provider-azure/blob/dont-share-vnet-name/api/v1beta1/types_class.go#L97

We can't fix this due to it being a breaking API change, so we just have a webhook warning for this instead: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4735/files#diff-18af7d1a5d1176aac1e0766c817b6e9d5e079a09e4605ff7c420a4d24d56f85cR43

Discussion about this is here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4735#discussion_r1571256564

File: api/v1beta1/types_class.go
Comment (jackfrancis) [reply api/v1beta1/types_class.go:418]: I don't fully understand the ramifications, let's maybe have a real-time discussion on a public zoom soon so we can figure this out

Comment (willie-yao) [reply api/v1beta1/types_class.go:418]: @jackfrancis We discussed this in last week's office hours and the general consensus was to move this out of the template type, but clarify in the release notes that this is a breaking change for Go API but not real-world usage.

Comment (jackfrancis) [reply api/v1beta1/types_class.go:418]: in which case lgtm

**Review by nojnhuh — APPROVED** [review]
/lgtm
/approve

**Comment by willie-yao** [comment]
Squashed! Messed up the force push so it'll need another lgtm

**Comment by nojnhuh** [comment]
Looks like what you have now matches what got LGTMs. Thanks!

/lgtm

**Pull request merged**
